### PR TITLE
SNOW-1057812: Support Open Rowset API Channel

### PR DESCRIPTION
**Pull request by sfc-gh-tzhang**
This PR contains the change to allow creation of a Rowset API channel:
- No behavior change on existing CLOUD_STORAGE interface
- Add an option to existing OpenChannelRequest to allow the creation of a Rowset API channel with ChannelType=ROWSET_API
- The OpenChannelRequest will be converted internally to a new REST API added for Rowset API, and we have a new implementation of the channel class (SnowflakeStreamingIngestChannelRowset) which is much simpler than the one for CLOUD_STORAGE

Note that we're merging into a feature branch instead of master

### CODE REVIEW

File: src/main/java/net/snowflake/ingest/streaming/SnowflakeStreamingIngestChannel.java [hunk 1]
```diff
@@ -247,6 +262,8 @@ public interface SnowflakeStreamingIngestChannel {
    */
   InsertValidationResponse insertRow(Map<String, Object> row, @Nullable String offsetToken);
 
+  InsertValidationResponse insertRow(Map<String, Object> row);
```
Comment (sfc-gh-asen) [inline src/main/java/net/snowflake/ingest/streaming/SnowflakeStreamingIngestChannel.java:265]: in which cases will a customers use insertRow without an offsetToken?

File: src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestChannelRowset.java [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+            "%s.%s.%s.%s", dbName, schemaName, pipeName == null ? tableName : pipeName, name);
+    this.offsetToken = offsetToken;
+    this.continuationToken = continuationToken;
+  }
+
+  /**
+   * Get the fully qualified channel name
+   *
+   * @return fully qualified name of the channel, in the format of
+   *     dbName.schemaName.tableName.channelName
```
Comment (sfc-gh-asen) [inline src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestChannelRowset.java:64]: nit : or dbName.schemaName.pipeName.channelName

File: src/main/java/net/snowflake/ingest/streaming/SnowflakeStreamingIngestChannel.java [hunk 3]
```diff
@@ -271,6 +288,8 @@ InsertValidationResponse insertRows(
   InsertValidationResponse insertRows(
       Iterable<Map<String, Object>> rows, @Nullable String offsetToken);
```
Comment (sfc-gh-asen) [inline src/main/java/net/snowflake/ingest/streaming/SnowflakeStreamingIngestChannel.java:289]: we might need to add something like

 insertRows(Iterable<Map<String, Object>> rows,  List<String> offsetTokens)
 
 for partial ingestion

File: src/main/java/net/snowflake/ingest/utils/Utils.java [hunk 4]
```diff
@@ -88,26 +90,25 @@ private static Provider instantiateSecurityProvider() {
...
 
-  /**
-   * Assert when the value is null
-   *
-   * @param name
-   * @param value
-   * @throws SFException
-   */
+  /** Assert when all the Strings are either null or Empty */
+  public static void assertStringsNotNullOrEmpty(String name, Collection<String> values)
```
Comment (sfc-gh-asen) [inline src/main/java/net/snowflake/ingest/utils/Utils.java:101]: nit: when I first saw this method, I assumed this was asserting that all the strings in the collection should be not null, maybe we can rename it to something like assertAtLeastOneStringNotNullOrEmpty to be extra clear?

File: src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestChannelInternal.java [hunk 5]
```diff
@@ -441,6 +472,17 @@ public Map<String, ColumnProperties> getTableSchema() {
...
+   * Get the type of channel, please referring to {@link
+   * net.snowflake.ingest.streaming.OpenChannelRequest.ChannelType} for the supported channel type
+   *
+   * @return type of the channel
+   */
+  @Override
+  public OpenChannelRequest.ChannelType getType() {
+    return OpenChannelRequest.ChannelType.CLOUD_STORAGE;
+  }
+
```
Comment (sfc-gh-asen) [inline src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestChannelInternal.java:485]: Should we rename this class to SnowflakeStreamingIngestChannelCloud to be parallel with the rowset?

File: src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestClientInternal.java [hunk 6]
```diff
@@ -299,76 +305,162 @@ public SnowflakeStreamingIngestChannelInternal<?> openChannel(OpenChannelRequest
...
 
-      return channel;
-    } catch (IOException | IngestResponseException e) {
-      throw new SFException(e, ErrorCode.OPEN_CHANNEL_FAILURE, e.getMessage());
-    }
+    // Channel is now registered, add it to the in-memory channel pool
+    return SnowflakeStreamingIngestChannelFactory.<T>builder(request.getChannelName())
+        .setDBName(request.getDBName())
+        .setSchemaName(request.getSchemaName())
+        .setTableName(request.getTableName())
```
Comment (sfc-gh-asen) [inline src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestClientInternal.java:455]: we need to setPipeName as well

**Review by sfc-gh-asen — COMMENTED** [review]
Thanks for the changes

File: src/main/java/net/snowflake/ingest/streaming/OpenChannelRequest.java [hunk 7]
```diff
@@ -16,6 +17,11 @@ public enum OnErrorOption {
     // indexes. No data is ingested
   }
 
+  public enum ChannelType {
+    CLOUD_STORAGE,
+    ROWSET_API,
+  }
```
Comment (sfc-gh-tjones) [inline src/main/java/net/snowflake/ingest/streaming/OpenChannelRequest.java:23]: I think this exists already? If you look at `StreamingResource` there's a check somewhere that compares against an enum.

File: src/main/java/net/snowflake/ingest/streaming/OpenChannelRequest.java [hunk 8]
```diff
@@ -114,19 +135,23 @@ private OpenChannelRequest(OpenChannelRequestBuilder builder) {
     Utils.assertStringNotNullOrEmpty("channel name", builder.channelName);
     Utils.assertStringNotNullOrEmpty("database name", builder.dbName);
     Utils.assertStringNotNullOrEmpty("schema name", builder.schemaName);
-    Utils.assertStringNotNullOrEmpty("table name", builder.tableName);
+    Utils.assertStringsNotNullOrEmpty(
```
[the remark this answers is not in the document]
Comment (sfc-gh-tjones) [reply src/main/java/net/snowflake/ingest/streaming/OpenChannelRequest.java:138]: Yeah this is a good edge case check to add.

File: src/main/java/net/snowflake/ingest/streaming/SnowflakeStreamingIngestChannel.java (hunk 1 above)
Comment (sfc-gh-tjones) [reply src/main/java/net/snowflake/ingest/streaming/SnowflakeStreamingIngestChannel.java:265]: Yeah shouldn't this be ` InsertValidationResponse insertRow(Map<String, Object> row, @Nullable String offsetToken);`?

File: src/main/java/net/snowflake/ingest/streaming/OpenChannelRequest.java (hunk 7 above)
Comment (sfc-gh-tzhang) [reply src/main/java/net/snowflake/ingest/streaming/OpenChannelRequest.java:23]: Yes, but the one used before was hided in the SDK, now this is a public enum in the public class

File: src/main/java/net/snowflake/ingest/streaming/OpenChannelRequest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (sfc-gh-tzhang) [reply src/main/java/net/snowflake/ingest/streaming/OpenChannelRequest.java:138]: The behavior right now is that the table name will be ignored if there is a pipe name, I thought about assert if both are provided but I think it's too restricted, it's perfect valid if customer inputs both and pipe is created on that table

File: src/main/java/net/snowflake/ingest/streaming/SnowflakeStreamingIngestChannel.java (hunk 1 above)
Comment (sfc-gh-tzhang) [reply src/main/java/net/snowflake/ingest/streaming/SnowflakeStreamingIngestChannel.java:265]: remember that we say the offset_token will be per row for Rowset API?

File: src/main/java/net/snowflake/ingest/streaming/SnowflakeStreamingIngestChannel.java (hunk 3 above)
Comment (sfc-gh-tzhang) [reply src/main/java/net/snowflake/ingest/streaming/SnowflakeStreamingIngestChannel.java:289]: Yes, that's the one below, offset_token will be part of the row now.

File: src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestChannelInternal.java (hunk 5 above)
Comment (sfc-gh-tzhang) [reply src/main/java/net/snowflake/ingest/streaming/internal/SnowflakeStreamingIngestChannelInternal.java:485]: Yes, I can do that in a separate PR

**Pull request closed without merging**